crates/sui-core/src/authority.rs

```diff
@@ -1098,42 +1100,63 @@ impl ExecutionState for AuthorityState {
+                // may input anything into consensus.
+                fp_ensure!(
+                    certificate.contains_shared_object(),
+                    SuiError::NotASharedObjectTransaction
+                );
+
+                // If we already executed this transaction, return the signed effects.
+                let digest = certificate.digest();
+                if self.database.effects_exists(digest)? {
+                    debug!(tx_digest =? digest, "Shared-object transaction already executed");
+                    let info = self.make_transaction_info(digest).await?;
+                    return Ok(bincode::serialize(&info).expect("Failed to serialize tx info"));
+                }
 
-        // Ensure the input is a shared object certificate. Remember that Byzantine authorities
-        // may input anything into consensus.
-        fp_ensure!(
-            certificate.contains_shared_object(),
-            SuiError::NotASharedObjectTransaction
-        );
+                // If we didn't already assigned shared-locks to this transaction, we do it now.
+                if !self.shared_locks_exist(&certificate).await? {
+                    // Check the certificate. Remember that Byzantine authorities may input anything into
+                    // consensus.
+                    certificate.verify(&self.committee)?;
+
+                    // Persist the certificate since we are about to lock one or more shared object.
+                    // We thus need to make sure someone (if not the client) can continue the protocol.
+                    // Also atomically lock the shared objects for this particular transaction and
+                    // increment the last consensus index. Note that a single process can ever call
+                    // this function and that the last consensus index is also kept in memory. It is
+                    // thus ok to only persist now (despite this function may have returned earlier).
+                    // In the worst case, the synchronizer of the consensus client will catch up.
+                    self.database.persist_certificate_and_lock_shared_objects(
+                        *certificate,
+                        execution_indices,
+                    )?;
+                }
 
-        // If we already executed this transaction, return the signed effects.
-        let digest = certificate.digest();
-        if self.database.effects_exists(digest)? {
-            debug!(tx_digest =? digest, "Shared-object transaction already executed");
-            let info = self.make_transaction_info(digest).await?;
-            return Ok(bincode::serialize(&info).expect("Failed to serialize tx info"));
-        }
+                // TODO: This return time is not ideal.
+                Ok(Vec::default())
+            }
+            ConsensusTransaction::Checkpoint(fragment) => {
+                let seq = execution_indices;
```

Comment:
I am a little confused about this execution index. This is the sequence in the stream right, not the version of objects executed.

Reply:
It is the sequence in the consensus (not Sui execution). The name is confusing indeed. It has to do with Narwhal terminology that composes badly with Sui. Narwhal has an execution engine and we are calling it to assign shared-locks (rather than execute Sui transactions). I can rename it in the trait though.